File: typescript/src/modules/town.ts
/**
 * 城镇管理服务模块
 * 基于 @sttot/axios-api 的标准化API定义
 */

import { apiBase } from '@sttot/axios-api';
import {
  Town,
  CreateTownRequest,
  GetTownRequest,
  UpdateTownRequest,
  DeleteTownRequest,
  ListTownsRequest,
  ListTownsResponse,
  GetTownMembersRequest,
  TownMembersResponse,
  TownDetailResponse,
  EmptyResponse
} from '../types';
import { commonErrorHandler } from '../utils/errorHandler';

export const initTownService = (apiFactory: ReturnType<typeof apiBase>) => {
  class TownService {

    // 创建城镇
    public createTown = apiFactory<CreateTownRequest, Town>(
      (request) => ({
        method: 'POST',
        url: '/api/v1/towns',
        data: {
          name: request.name,
          level: request.level,
          leader_id: request.leaderId,
          qq_group: request.qqGroup,
          description: request.description
        }
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        level: data.level,
        leaderId: data.leader_id,
        qqGroup: data.qq_group,
        description: data.description,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

        // 获取城镇列表
    public listTowns = apiFactory<ListTownsRequest, ListTownsResponse>(
      (request = {}) => ({
        method: 'GET',
        url: '/api/v1/towns',
        params: this.buildParams({
          page: request.page,
          page_size: request.pageSize,
          name: request.name,
          search: request.search,
          min_level: request.minLevel,
          max_level: request.maxLevel
        })
      }),
      ({ data }) => ({
        towns: data.towns.map((town: any) => ({
          id: town.id,
          name: town.name,
          level: town.level,
          leaderId: town.leader_id,
          qqGroup: town.qq_group,
          description: town.description,
          createdAt: town.created_at,
          updatedAt: town.updated_at
        })),
        total: data.total,
        page: data.page,
        pageSize: data.page_size
      }),
      commonErrorHandler
    );

    // 获取城镇详情（根据IDL规范，返回TownDetailResponse）
    public getTown = apiFactory<GetTownRequest, TownDetailResponse>(
      (request) => ({
        method: 'GET',
        url: `/api/v1/towns/${request.id}`,
        params: request.detail ? { detail: request.detail } : undefined
      }),
      ({ data }) => ({
        town: {
          id: data.town.id,
          name: data.town.name,
          level: data.town.level,
          leaderId: data.town.leader_id,
          qqGroup: data.town.qq_group,
          description: data.town.description,
          createdAt: data.town.created_at,
          updatedAt: data.town.updated_at
        },
        leader: data.leader,  // 直接使用ID，符合IDL中的 optional i32 leader 定义
        members: data.members.map((member: any) => ({
          id: member.id,
          name: member.name,
          townId: member.town_id,
          qq: member.qq,
          qqguild: member.qqguild,
          discord: member.discord,
          inQqGroup: member.in_qq_group,
          inQqGuild: member.in_qq_guild,
          inDiscord: member.in_discord,
          banMode: member.ban_mode,
          banExpire: member.ban_expire,
          banReason: member.ban_reason,
          createdAt: member.created_at,
          updatedAt: member.updated_at
        }))
      }),
      commonErrorHandler
    );

    // 更新城镇信息
    public updateTown = apiFactory<UpdateTownRequest, Town>(
      (request) => ({
        method: 'PUT',
        url: `/api/v1/towns/${request.id}`,
        data: {
          name: request.name,
          level: request.level,
          leader_id: request.leaderId,
          qq_group: request.qqGroup,
          description: request.description
        }
      }),
      ({ data }) => ({
        id: data.id,
        name: data.name,
        level: data.level,
        leaderId: data.leader_id,
        qqGroup: data.qq_group,
        description: data.description,
        createdAt: data.created_at,
        updatedAt: data.updated_at
      }),
      commonErrorHandler
    );

    // 删除城镇
    public deleteTown = apiFactory<DeleteTownRequest, EmptyResponse>(
      (request) => ({
        method: 'DELETE',
        url: `/api/v1/towns/${request.id}`
      }),
      ({ data }) => data as EmptyResponse,
      commonErrorHandler
    );

    // 获取城镇成员列表
    public getTownMembers = apiFactory<GetTownMembersRequest, TownMembersResponse>(
      (request) => ({
        method: 'GET',
        url: `/api/v1/towns/${request.townId}/members`,
        params: this.buildParams({
          page: request.page,
          pageSize: request.pageSize  // 使用符合IDL规范的字段命名
        })
      }),
      ({ data }) => ({
        members: data.members.map((member: any) => ({
          playerId: member.player_id,
          playerName: member.player_name,
          joinedAt: member.joined_at,
          role: member.role
        })),
        total: data.total,
        page: data.page,
        pageSize: data.page_size
      }),
      commonErrorHandler
    );



    // 辅助方法：构建查询参数，过滤掉undefined值
    public buildParams(params: Record<string, any>): Record<string, any> {
      const result: Record<string, any> = {};
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          result[key] = value;
        }
      }
      return result;
    }
  }

  return new TownService();
};

File: golang/modules/town.go
package modules

import (
	"strconv"

	"github.com/Gk0Wk/NewNanManager/sdk/golang/utils"
	"github.com/go-resty/resty/v2"
)

// TownService 城镇管理服务
type TownService struct {
	client *resty.Client
}

// NewTownService 创建城镇服务实例
func NewTownService(client *resty.Client) *TownService {
	return &TownService{client: client}
}

// ListTowns 获取城镇列表
func (s *TownService) ListTowns(page, pageSize *int32, search *string, minLevel, maxLevel *int32) (*TownsListData, error) {
	req := s.client.R()

	if page != nil {
		req.SetQueryParam("page", strconv.Itoa(int(*page)))
	}
	if pageSize != nil {
		req.SetQueryParam("page_size", strconv.Itoa(int(*pageSize)))
	}
	if search != nil {
		req.SetQueryParam("search", *search)
	}
	if minLevel != nil {
		req.SetQueryParam("min_level", strconv.Itoa(int(*minLevel)))
	}
	if maxLevel != nil {
		req.SetQueryParam("max_level", strconv.Itoa(int(*maxLevel)))
	}

	resp, err := req.Get("/api/v1/towns")

	var result TownsListData
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateTown 创建城镇
func (s *TownService) CreateTown(request CreateTownRequest) (*Town, error) {
	resp, err := s.client.R().
		SetBody(request).
		Post("/api/v1/towns")

	var result Town
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// GetTown 获取城镇详情
func (s *TownService) GetTown(id int32, detail bool) (*TownDetailResponse, error) {
	req := s.client.R()

	if detail {
		req.SetQueryParam("detail", "true")
	}

	resp, err := req.Get("/api/v1/towns/" + strconv.Itoa(int(id)))

	var result TownDetailResponse
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdateTown 更新城镇信息
func (s *TownService) UpdateTown(id int32, request UpdateTownRequest) (*Town, error) {
	resp, err := s.client.R().
		SetBody(request).
		Put("/api/v1/towns/" + strconv.Itoa(int(id)))

	var result Town
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// DeleteTown 删除城镇
func (s *TownService) DeleteTown(id int32) error {
	resp, err := s.client.R().
		Delete("/api/v1/towns/" + strconv.Itoa(int(id)))

	return utils.HandleResponse(resp, err, nil)
}

// GetTownMembers 获取城镇成员列表
func (s *TownService) GetTownMembers(townID int32, page, pageSize *int32) (*TownMembersData, error) {
	req := s.client.R()

	if page != nil {
		req.SetQueryParam("page", strconv.Itoa(int(*page)))
	}
	if pageSize != nil {
		req.SetQueryParam("page_size", strconv.Itoa(int(*pageSize)))
	}

	resp, err := req.Get("/api/v1/towns/" + strconv.Itoa(int(townID)) + "/members")

	var result TownMembersData
	err = utils.HandleResponse(resp, err, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}
